### PR TITLE
Fix BatchNormalization training output depending on momentum

### LINKED ISSUE #29790: [BUG] Output Y from BatchNormalization depends on momentum when training_mode=1
### Describe the issue

Per the [documentation](https://onnx.ai/onnx/operators/onnx__BatchNormalization.html), when `training_mode=1`, the `Y` output from BatchNormalization is given by
```
Y = (X - current_mean) / sqrt(current_var + epsilon) * scale + B
```
That is, `Y` is not supposed to depend on `input_mean`, `input_var`, or `momentum`.

However, in my testing, `Y` does change depending on `momentum` but not in a (to me) sensible way. The output is wrong unless I set `momentum=1`; then `Y` matches the definition above.

The output is correct when `training_mode=0`.

### To reproduce

I have created three ONNX models containing only a BatchNormalization operator with `scale=1`, `B=0`, `input_mean=0`, `input_var=1`, `training_mode=1`, and `momentum` taking on `0.0`, `0.5`, and `1.0`, respectively. There is a single float32 input of size 2x4x3.

Running these models with the same input should yield the same output, but this is not the case. I have tried both using the C API and the Julia bindings ONNXRunTime.jl.

[models.zip](https://github.com/user-attachments/files/30187895/models.zip)

### Urgency

No urgency.

### Platform

Linux

### OS Version

Ubuntu 22.04.1 LTS

### ONNX Runtime Installation

Released Package

### ONNX Runtime Version or Commit ID

1.27.1

### ONNX Runtime API

C

### Architecture

X64

### Execution Provider

Default CPU

### Execution Provider Library Version

_No response_

### PR DESCRIPTION
**Pull request by tairenpiao**
### Problem
With `training_mode=1`, `Y` changes with the `momentum` attribute and is only correct at
`momentum=1`. Per the spec `Y` comes from the current batch statistics and must not depend on
`momentum`. The error scales with `1 - momentum`.

### Root cause
The kernel holds Eigen *views* of `scale`/`B`, writes `running_mean`/`running_var`, then computes
`Y` from those views. Unused running outputs are backed by the input buffers, so when one
initializer feeds both `scale`/`B` and `input_mean`/`input_var` the write clobbers `scale`/`B`
first. `ConstantSharing` creates that aliasing by merging equal initializers:

    before: ['X', 'scale1',     'bias1',       'input_mean1', 'input_var1']
    after:  ['X', 'input_var1', 'input_mean1', 'input_mean1', 'input_var1']

A hand-written model is wrong at `ORT_DISABLE_ALL` too, so the fix belongs in the kernel.

### Fix
Write the running statistics after `Y`, when nothing reads `scale`/`B`. They are captured before
`saved_inv_std` becomes `inv_std` in place, so `running_var` still gets the batch variance;
inference is unaffected. The test runs at `TransformerLevel::Default` (`Y` is off by 0.84
unpatched). I couldn't link the test binary locally — unrelated protobuf conflict — so relying
on CI.

Fixes #29790

### CODE REVIEW

**Comment by tairenpiao** [comment]
Hi @tianleiwu, could anyone review this PR?

File: onnxruntime/core/providers/cpu/nn/batch_norm.h [hunk 1]
```diff
@@ -203,6 +205,15 @@ class BatchNorm : public OpKernel {
         Y_arr.col(n) = X_arr.col(n) * new_scale.col(0) + new_bias.col(0);
       }
     }
+
+#if defined(BATCHNORM_INCLUDE_TRAINING_SUPPORT)
+    // Safe to write now that Y no longer reads scale/B.
+    if (is_train_) {
+      EigenVectorArrayMap<T>(running_mean->MutableData<T>(), C) = new_running_mean;
```
Comment (tianleiwu) [inline onnxruntime/core/providers/cpu/nn/batch_norm.h]: This still writes through the mandatory input/output alias, so the fix only holds for the first `Run`. In the added model, this changes the shared zero initializer (also `B`) to the batch mean `0.5`, and the shared one initializer (also `scale`) to the batch variance `1.25`. A second `session.Run` with the same input then uses `B = 0.5` and `scale = 1.25`, producing approximately `[-1.1770, -0.0590, 1.0590, 2.1770]` instead of the expected normalized values. Please ensure these outputs do not write through storage shared with `scale`/`B` (the mandatory alias likely needs to be removed or made conditional), and run the new regression session at least twice to cover this state transition.

**Review by tianleiwu — CHANGES_REQUESTED** [review]
Deferring the running-statistic writes fixes the first invocation, but the shared initializer buffers are still mutated afterward. The same session therefore produces incorrect `Y` on subsequent runs; details are inline.

File: onnxruntime/core/providers/cpu/nn/batch_norm.h (hunk 1 above)
Comment (tairenpiao) [reply onnxruntime/core/providers/cpu/nn/batch_norm.h]: Updated. Please take look.

**Comment by martinkjlarsson** [comment]
What's the status of this PR? Is it just waiting for review?

**Comment by tairenpiao** [comment]
Hi @martinkjlarsson, yes, it's waiting for a re-review. The 8/4 feedback was addressed on 8/14: the running mean/var outputs no longer alias the inputs, and the regression test now runs the session twice. @tianleiwu could you take another look and approve the CI runs when you get a chance? Thanks!
